### PR TITLE
* download sources property propagation. relates to #372

### PR DESCRIPTION
**Pull request by dmitry-zhuravlev**

### CODE REVIEW

File: modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/maven/aether/KobaltMavenResolver.kt [hunk 1]
```diff
@@ -38,6 +38,14 @@ class KobaltMavenResolver @Inject constructor(val settings: KobaltSettings,
     fun resolve(id: String, scope: Scope? = null, filter: DependencyFilter? = null): DependencyResult {
         val dependencyRequest = DependencyRequest(createCollectRequest(id, scope), filter)
         val result = system.resolveDependencies(session, dependencyRequest)
+        if (args.downloadSources) {
+            listOf("sources", "javadoc").forEach {
+                val artifact = DefaultArtifact(id)
+                val sourceArtifact = DefaultArtifact(artifact.groupId, artifact.artifactId, it, artifact.extension,
+                        artifact.version)
+                system.resolveDependencies(session, DependencyRequest(createCollectRequest(sourceArtifact, scope), filter))
```
Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/maven/aether/KobaltMavenResolver.kt]: Why this change? By doing this in `createCollectRequest()`, I actually only send one resolution to MavenResolver, and it will download everything in one go. Here, you are doing three separate `resolveDependencies` (one for the regular jar, one for the sources and one for the javadocs).

Comment (dmitry-zhuravlev) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/maven/aether/KobaltMavenResolver.kt]: I just want it to separate from the regular jar  dependencies, otherwise you collect sources and javadoc like a standalone dependencies and send it all to plugin. Feel free to refactor it if needed.

Comment (dmitry-zhuravlev) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/maven/aether/KobaltMavenResolver.kt]: In case if you want exactly one resolution in MavenResolver you should somehow filter session.resolveDependencies(...) DependencyResult and exclude sources/javadoc jars. So probably several separate resolutions in MavenResolver make sense....

Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/maven/aether/KobaltMavenResolver.kt]: I just experimented with launching the server with `--downloadSources` and then synced, and it didn't go well. 

First of all, it downloads all javadoc/sources for all the dependencies (which is terrible) but worse, if one of these dependencies fails to resolve (e.g. javadoc missing, which happens quite a bit on JCenter), the whole process aborts.

So this is the wrong solution: the plug-in should never launch the server in `--downloadSources` mode.

Instead, let's implement the "Download sources" solution as I suggested. Maybe when the user presses this button, if you can't resolve this yourself, you can either

- Launch a second Kobalt process with `--downloadSources --resolve ...`
- Or if you don't want to do that, I can add an API call to do this and you just call the server with that API call to download the sources.

What do you think?

**Comment by dmitry-zhuravlev** [comment]
>but worse, if one of these dependencies fails to resolve (e.g. javadoc missing, which happens quite a bit on JCenter), the whole process aborts.

Why it fails? Especially in case if resolution of sources/javadoc processed in separated session.resolveDependencies(...) calls? Maybe there need to add additional exeption catch blocks, i'll take a look.

Regarding "Download Sources" button: it is a build-in ability in IDEA and should work fine (except some bug with HTTP proxy). I don't think that we need to reinvent this button again with Kobalt. 

In general this "Download sources" button and "Download All Sources for all dependencies" is a different features and should be present both.
Sometimes it is convinient to not download sources at all, somtimes it is convinient to download one particular source for one dependency, and sometimes it is a good to have all sources available without manually pressing "Download sources" button each time.

**Comment by cbeust** [comment]
Awesome!

I made a couple of refactorings on top of your PR, can you take a look?

- https://github.com/cbeust/kobalt/commit/bb2eb19876c89973efdf45a9713cdb94212ac53d
- https://github.com/cbeust/kobalt/commit/0f7bb021c38eb0121b682cb20e53220ae6e6313e

**Comment by cbeust** [comment]
This PR is now deployed.

**Comment by dmitry-zhuravlev** [comment]
Thanks for refactoring! Looks good!
